File: c/medium/maximize_area_of_square_hole_in_grid.c
#include <stdlib.h>

int cmp_int(const void *a, const void *b) {
    return (*(int *)a) - (*(int *)b);
}

int longest_consecutive_run(int *bars, int size) {
    if (size == 0) return 0;

    int max_run = 1;
    int current_run = 1;

    for (int i = 1; i < size; i++) {
        if (bars[i] == bars[i - 1] + 1) {
            current_run++;
        } else {
            if (current_run > max_run)
                max_run = current_run;
            current_run = 1;
        }
    }

    return current_run > max_run ? current_run : max_run;
}

int maximizeSquareHoleArea(int n, int m, int *hBars, int hSize, int *vBars, int vSize) {
    qsort(hBars, hSize, sizeof(int), cmp_int);
    qsort(vBars, vSize, sizeof(int), cmp_int);

    int max_horizontal_run = longest_consecutive_run(hBars, hSize);
    int max_vertical_run = longest_consecutive_run(vBars, vSize);

    int max_square_side = (max_horizontal_run < max_vertical_run
                           ? max_horizontal_run
                           : max_vertical_run) + 1;

    return max_square_side * max_square_side;
}

File: rust/medium/maximize_area_of_square_hole_in_grid.rs
fn longest_consecutive_run(bars: &Vec<i32>) -> i32 {
    if bars.is_empty() {
        return 0;
    }

    let mut max_run = 1;
    let mut current_run = 1;

    for i in 1..bars.len() {
        if bars[i] == bars[i - 1] + 1 {
            current_run += 1;
        } else {
            max_run = max_run.max(current_run);
            current_run = 1;
        }
    }

    max_run.max(current_run)
}

fn maximize_square_hole_area(
    _n: i32,
    _m: i32,
    mut h_bars: Vec<i32>,
    mut v_bars: Vec<i32>,
) -> i32 {
    h_bars.sort();
    v_bars.sort();

    let max_horizontal_run = longest_consecutive_run(&h_bars);
    let max_vertical_run = longest_consecutive_run(&v_bars);

    let max_square_side = std::cmp::min(max_horizontal_run, max_vertical_run) + 1;
    max_square_side * max_square_side
}
